src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,159 @@
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexInBoundsStrict(index);
+        if (index == size) {
+            add(value);
+        } else {
+            linkBefore(value, getNodeByIndex(index));
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexInBoundsNotStrict(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexInBoundsStrict(index);
+        Node<T> node = getNodeByIndex(index);
+        T previousValue = node.value;
+        node.value = value;
+        return previousValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexInBoundsNotStrict(index);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> i = first; i != null; i = i.next) {
+            if (i.value == object || (object != null && object.equals(i.value))) {
+                unlink(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        indexInBoundsNotStrict(index);
+        Node<T> node;
+        if (index < (size / 2)) {
+            node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.previous;
+            }
+        }
+        return node;
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> nextNode = node.next;
+        Node<T> previousNode = node.previous;
+        if (previousNode == null) {
+            first = nextNode;
+        } else {
+            previousNode.next = nextNode;
+            node.previous = null;
+        }
+        if (nextNode == null) {
+            last = previousNode;
+        } else {
+            nextNode.previous = previousNode;
+            node.next = null;
+        }
+        T removableNode = node.value;
+        node.value = null;
+        size--;
+        return removableNode;
+    }
+
+    private void indexInBoundsNotStrict(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private void indexInBoundsStrict(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
```

Comment:
In my opinion here is good to do constant INDEX_OUT_OF_BOUND_MESSAGE = ""Index out of bounds..."